### PR TITLE
glib: add a Value::get_owned

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
The current FromValue trait can't handle the FromValue<'static>
case as that means having to borrow self for 'static.

Needed for having a generic ObjectExt::property that would return T instead of Value. See #323

### CODE REVIEW

File: glib/src/value.rs [hunk 1]
```diff
@@ -205,6 +205,18 @@ pub unsafe trait FromValue<'a>: Sized {
     unsafe fn from_value(value: &'a Value) -> Self;
 }
 
+/// Similar to `FromValue` but works for owned `Value.
```
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply glib/src/value.rs:208]: it's hard to do because the `ToValue` trait is not implemented for `&'static str`

[the remark this answers is not in the document]
Comment (sdroege) [reply glib/src/value.rs:208]: Why not? It really should, that's how `obj.set_property("foo", "bar")` works :)

File: glib/src/value.rs [hunk 2]
```diff
@@ -205,6 +205,18 @@ pub unsafe trait FromValue<'a>: Sized {
...
+pub unsafe trait FromOwnedValue: Sized {
```
Comment (sdroege) [inline glib/src/value.rs:209]:
```suggestion
pub unsafe trait FromOwnedValue: for<'a> FromValue<'a> {
```

Then you can remove the `Checker` probably

File: glib/src/value.rs [hunk 3]
```diff
@@ -205,6 +205,18 @@ pub unsafe trait FromValue<'a>: Sized {
...
+    /// Value type checker.
+    type Checker: ValueTypeChecker;
+
+    /// Get the contained value from a `Value`
+    ///
+    /// # Safety
+    /// `Self::Checker::check()` must be called first and must not fail.
+    unsafe fn from_value(value: &Value) -> Self;
```
Comment (sdroege) [inline glib/src/value.rs:217]:
```suggestion
    unsafe fn from_value_owned(value: &Value) -> Self;
```

File: glib/src/value.rs [hunk 4]
```diff
@@ -205,6 +205,18 @@ pub unsafe trait FromValue<'a>: Sized {
     unsafe fn from_value(value: &'a Value) -> Self;
```
Comment (sdroege) [inline glib/src/value.rs:205]: Search for `FromValue<'static>` in this file. These should all become `FromValueOwned` AFAICS

File: glib/src/value.rs [hunk 5]
```diff
@@ -205,6 +205,18 @@ pub unsafe trait FromValue<'a>: Sized {
...
+    /// Value type checker.
+    type Checker: ValueTypeChecker;
+
+    /// Get the contained value from a `Value`
+    ///
+    /// # Safety
+    /// `Self::Checker::check()` must be called first and must not fail.
+    unsafe fn from_value(value: &Value) -> Self;
+}
+
```
Comment (sdroege) [inline glib/src/value.rs:219]: You'll actually have to implement the new trait for `String`, `i32`, etc. (and also put it into the `glib::wrapper!` macros for boxed/shared/boxed-inline/object)

File: glib/src/value.rs (hunk 4 above)
Comment (bilelmoussaoui) [reply glib/src/value.rs:205]: All of them? the one in ValueType trait definition can't be made as FromValueOwned, it breaks a ton of code

Comment (sdroege) [reply glib/src/value.rs:205]: All two of them, and especially that one was the one I was thinking of. How does it break?

File: glib/src/value.rs [hunk 6]
```diff
@@ -205,6 +205,21 @@ pub unsafe trait FromValue<'a>: Sized {
...
+/// let x = "gtk-rs".to_value();
+///
+/// let y = x.get_owned::<&str>();
+/// ```
+pub unsafe trait FromOwnedValue: for<'a> FromValue<'a> {
+    /// Get the contained value from a `Value`
+    ///
+    /// # Safety
+    /// `Self::Checker::check()` must be called first and must not fail.
+    unsafe fn from_value_owned(value: &Value) -> Self;
```
Comment (sdroege) [inline glib/src/value.rs:220]:
```suggestion
    unsafe fn from_value_owned(value: &Value) -> Self {
        <Self as FromValue>::from_value(v)
    }
```

Does that work? Then you don't need to implement any functions anywhere

File: glib/src/value.rs [hunk 7]
```diff
@@ -964,6 +1038,7 @@ mod tests {
 
         let some_v = Some("test").to_value();
         assert_eq!(some_v.get::<&str>(), Ok("test"));
+        assert_eq!(some_v.get_owned::<String>(), Ok("test".to_string()));
```
Comment (sdroege) [inline glib/src/value.rs:1041]: Can you also add a test for `Option<String>`?

File: glib/src/value.rs [hunk 8]
```diff
@@ -783,6 +851,12 @@ macro_rules! numeric {
             type Checker = GenericValueTypeChecker<Self>;
 
             unsafe fn from_value(value: &'a Value) -> Self {
+                FromOwnedValue::from_value_owned(value)
+            }
+        }
+
+        unsafe impl FromOwnedValue for $name {
```
Comment (sdroege) [inline glib/src/value.rs:858]: Please also shortly grep in the sources to see if there are any manual `FromValue` impls left that should get this one too

File: glib/src/value.rs [hunk 9]
```diff
@@ -205,6 +205,21 @@ pub unsafe trait FromValue<'a>: Sized {
...
 }
 
+/// Similar to `FromValue` but works for owned `Value.
+///
+/// ```compile_fail
+/// let x = "gtk-rs".to_value();
+///
+/// let y = x.get_owned::<&str>();
+/// ```
+pub unsafe trait FromOwnedValue: for<'a> FromValue<'a> {
```
Comment (sdroege) [inline glib/src/value.rs:215]:
```suggestion
pub unsafe trait FromOwnedValue: for<'a> FromValue<'a> + 'static {
```

File: glib/src/value.rs [hunk 10]
```diff
@@ -228,6 +243,15 @@ mod private {
     }
 }
 
+pub trait FromOwnedValueOptional: for<'a> FromValueOptional<'a> {}
```
Comment (sdroege) [inline glib/src/value.rs:246]:
```suggestion
pub trait FromOwnedValueOptional: for<'a> FromValueOptional<'a> + 'static {}
```

File: glib/src/value.rs (hunk 6 above)
Comment (bilelmoussaoui) [reply glib/src/value.rs:220]: Indeed, thanks!

File: glib/src/value.rs (hunk 8 above)
Comment (bilelmoussaoui) [reply glib/src/value.rs:858]: the only one left are Value & BoxedValue. Should they also get a FromOwnedValue impl?

File: glib/src/value.rs (hunk 4 above)
Comment (bilelmoussaoui) [reply glib/src/value.rs:205]: There are different cases:
1 - We need to generate the new trait as part of gir & do a regen (for flags/enums)
2 - we can no longer impl ValueType for `&'static str`
3 - BoxedValue 

Otherwise I fixed the remaining ones

Comment (bilelmoussaoui) [reply glib/src/value.rs:205]: For 1- https://github.com/gtk-rs/gir/pull/1254

File: glib/src/value.rs (hunk 8 above)
Comment (sdroege) [reply glib/src/value.rs]: Yes

File: glib/src/value.rs (hunk 4 above)
Comment (sdroege) [reply glib/src/value.rs:205]: 2 - seems expected and correct
3 - what's the problem there?

Comment (bilelmoussaoui) [reply glib/src/value.rs:205]: 3 - None, i did implement FromOwnedValue for it as well
2 - so, we drop the ValueType impl of it?

**Comment by bilelmoussaoui** [comment]
This one is blocked by https://github.com/gtk-rs/gir/pull/1254 now

File: glib/src/value.rs (hunk 4 above)
Comment (sdroege) [reply glib/src/value.rs:205]: 2 - Yes, it makes no sense. You can't get a `&'static str` from a `Value`

File: glib/src/value.rs [hunk 11]
```diff
@@ -205,6 +205,23 @@ pub unsafe trait FromValue<'a>: Sized {
     unsafe fn from_value(value: &'a Value) -> Self;
 }
 
+/// Similar to `FromValue` but works for owned `Value.
+///
+/// ```compile_fail
+/// let x = "gtk-rs".to_value();
```
Comment (GuillaumeGomez) [inline glib/src/value.rs:211]: Maybe add some comments like "This example cannot compile because `&str` is a temporary value" or something along the line. And please also add a working code example (with `String` for example). So both are always checked.

**Review by GuillaumeGomez — APPROVED** [review]

File: glib/src/value.rs [hunk 12]
```diff
@@ -205,6 +205,29 @@ pub unsafe trait FromValue<'a>: Sized {
...
+///
+/// let y = x.get_owned::<&str>();
+/// ```
+/// 
+/// ```
+/// let owned_str = "hello".to_string().to_value();
+/// assert_eq!(owned_str.owned_get::<String>(), "hello".to_string());
+/// ```
+pub unsafe trait FromOwnedValue: for<'a> FromValue<'a> + 'static {
```
Comment (sdroege) [inline glib/src/value.rs:221]: I wonder if we can make a blanket implementation for this instead of implementing it for every type. Does something like

```rust
unsafe impl<'a, T: FromValue<'a> + 'static> FromOwnedValue for T {}
```

work?

Comment (bilelmoussaoui) [reply glib/src/value.rs:221]: That can't work because we don't have a FromValue impl for String. If we manually implement FromOwnedValue for String only, it complains that it's already implemented as part of the blanket implementation :/

File: glib/src/object.rs [hunk 13]
```diff
@@ -45,7 +45,7 @@ pub unsafe trait ObjectType:
     + crate::value::ValueType
     + crate::value::ToValue
     + crate::value::ToValueOptional
-    + crate::value::FromValueOptional<'static>
+    + crate::value::FromOwnedValueOptional
```
Comment (sdroege) [inline glib/src/object.rs:48]:
```suggestion
    + crate::value::FromOwnedValueOptional
    + for<'a> crate::value::FromValueOptional<'a>
```

too maybe?

File: glib/src/value.rs [hunk 14]
```diff
@@ -395,6 +432,17 @@ impl Value {
         }
     }
 
+    /// Tries to get a value of an owned type `T`.
+    pub fn get_owned<T>(&self) -> Result<T, <<T as FromValue>::Checker as ValueTypeChecker>::Error>
+    where
+        T: FromOwnedValue,
```
Comment (sdroege) [inline glib/src/value.rs:438]:
```suggestion
        T: for<'b> FromValue<'b> + 'static,
```

Does this maybe also work?

File: glib/src/param_spec.rs [hunk 15]
```diff
@@ -43,7 +43,7 @@ impl crate::value::ValueType for ParamSpec {
 unsafe impl<'a> crate::value::FromValue<'a> for ParamSpec {
     type Checker = crate::value::GenericValueTypeOrNoneChecker<Self>;
 
-    unsafe fn from_value(value: &'a crate::Value) -> Self {
+    unsafe fn from_value(value: &crate::Value) -> Self {
```
Comment (sdroege) [inline glib/src/param_spec.rs:46]: Why this change, and why does this even compile? :)

Comment (bilelmoussaoui) [reply glib/src/param_spec.rs:46]: Heh, not an intended change.

File: glib/src/value.rs (hunk 14 above)
Comment (sdroege) [reply glib/src/value.rs:438]: If that works we won't need a new trait, if it doesn't then that's just how it is. With the blanket implementation it should be good enough.

File: glib/src/value.rs [hunk 16]
```diff
@@ -52,7 +52,7 @@ use crate::translate::*;
 use crate::types::{StaticType, Type};
 
 /// A type that can be stored in `Value`s.
-pub trait ValueType: ToValue + FromValue<'static> + 'static {
+pub trait ValueType: ToValue + FromValue<'static> {
```
Comment (sdroege) [inline glib/src/value.rs:55]:
```suggestion
pub trait ValueType: ToValue + for<'a> FromValue<'a> + 'static {
```

Comment (sdroege) [reply glib/src/value.rs:55]: Also the other `FromValue<'static>` cases in this file

**Review by sdroege — APPROVED** [review]
